### ISSUE TITLE
[Eval] Implement post-training evaluation suite for paper submission

### ISSUE THREAD

**Issue by Aviral-03** [UofT-CSC490-W2026/T-MoE#22]
## Context & What Is Already Implemented

Before contributing, understand what is **already handled by `scripts/train.py`** so you do not duplicate it:

| Metric | Status | WandB Key | Frequency |
|---|---|---|---|
| Training loss + PPL | ✅ Done | `train/loss`, `train/ppl` | every `log_interval` steps |
| Validation loss + PPL | ✅ Done | `val_loss`, `val_ppl` | every `eval_interval` steps |
| Gini coefficient | ✅ Done | `router/{layer}/routing_diversity_gini` | every `log_interval` |
| Effective experts | ✅ Done | `router/{layer}/effective_experts` | every `log_interval` |
| Per-expert usage | ✅ Done | `router/{layer}/expert_{i}_usage` | every `log_interval` |
| Fatigue stats | ✅ Done | `router/{layer}/fatigue_{mean,std,min,max}` | every `log_interval` |
| Specialization / MI | ✅ Done | `specialization/{tracker}/specialization_score` etc. | every `log_interval` |
| Training throughput | ✅ Done | `train/tok_per_sec`, `train/dt` | every `log_interval` |

**Do not re-implement any of the above.** This issue is exclusively about **post-training evaluation** that runs on a saved checkpoint after the training job completes.

---

## Data & Training Flow

Understanding the pipeline is required before writing any eval code:

```
scripts/prepare_data.py
  └─ Downloads dataset (HF Hub, streaming for large sets)
  └─ Tokenizes + packs into binary shards
  └─ Output: data/shards/{dataset_key}/{split}_shard_{idx:04d}.bin
       Format: [uint64 token_count (8 bytes)] [uint16 tokens ...]

scripts/train.py
  └─ ShardDataset: memory-maps shard files (no full RAM load)
  └─ DistributedSampler for DDP ranks
  └─ Trains for N steps; saves checkpoints via CheckpointManager
       Output: outputs/{experiment_name}/checkpoint_step_{N}.pt
                outputs/{experiment_name}/best_model.pt
  └─ Logs training + routing metrics to WandB in real time

scripts/eval.py   ← THIS ISSUE
  └─ Loads a saved checkpoint (offline, single GPU, no DDP needed)
  └─ Runs downstream benchmarks, generation evals, inference profiling
  └─ Writes results to outputs/{experiment_name}/eval/{task}.json
  └─ Logs summary to WandB under project tmoe, run eval/{experiment_name}
```

**Eval always runs after training, on a single GPU, against a frozen checkpoint.** It must not require the same distributed setup as training.

---

### 1.1 Language Modeling — Standard Test Benchmarks

**Why:** `train.py` already logs val PPL on the training data distribution (wikitext-103 validation split, during training). What is **missing** is evaluation on the **held-out test split** and on **out-of-distribution corpora** — both required for a publishable perplexity table.

| Metric | Dataset | Split | Methodology |
|---|---|---|---|
| Bits-per-byte (BPB) | WikiText-103 | `test` | Sliding window, stride=512 |
| Perplexity | Penn Treebank | `test` | Full sequence |
| Perplexity | The Pile — Wikipedia + Books subsets | `validation` | Stride=512 |

**Notes:**
- Report BPB as primary metric — it is tokenizer-agnostic and directly comparable across model families.
- Use stride=512 sliding window. Do not use non-overlapping chunks — this underestimates model quality on long contexts.
- Run in `bf16`, matching training precision. Use `torch.no_grad()`.
- Both metabolic_v4 and standard_v2 checkpoints must be evaluated identically.

**Implementation:** `evals/perplexity.py`, invoked via `scripts/eval.py --task perplexity`

---

### 1.2 Downstream Task Benchmarks

**Why:** Val PPL on the training distribution is necessary but not sufficient. Reviewers will ask whether better routing quality translates to better downstream task performance. These benchmarks are held-out from training and test generalisation.

Use [`lm-evaluation-harness`](https://github.com/EleutherAI/lm-evaluation-harness) (EleutherAI) as the backend — do not reimplement tasks. This is the same framework used to evaluate Pythia, OPT, and LLaMA, enabling direct comparison.

| Task | Dataset | Metric | Shot | Notes |
|---|---|---|---|---|
| Commonsense NLI | **HellaSwag** | Acc (norm) | 0-shot | Standard for all LM papers |
| Physical reasoning | **PIQA** | Acc | 0-shot | Grounded world knowledge |
| Coreference | **WinoGrande** | Acc | 0-shot | Referential reasoning |
| Science QA | **ARC-Easy** | Acc (norm) | 0-shot | Well-calibrated at sub-3B |
| Science QA | **ARC-Challenge** | Acc (norm) | 0-shot | Model ceiling |
| Multi-task knowledge | **MMLU** | Acc | 5-shot | Required for any 2024+ submission |
| Math reasoning | **GSM8K** | Exact match | 8-shot CoT | 1.4B+ only; 125M will score near zero — report but do not highlight |

**Implementation:** `evals/lm_harness_runner.py` wrapping `lm_eval.simple_evaluate()`, invoked via `scripts/eval.py --task lm_harness`

---

### 1.3 Generation Quality (LLM-as-a-Judge)

**Why:** Multiple-choice accuracy does not capture open-ended generation quality. For a routing paper, this validates that expert specialisation (the MI layer-depth gradient already observed in training) produces semantically coherent outputs, not just better next-token log-likelihoods. Run this on final checkpoints only — not every eval interval.

**Methodology:**
1. **Prompt set:** 200 fixed prompts in `evals/judge_prompts.jsonl`, split across three domains — factual (70), reasoning (70), creative (60). **Prompts must be finalised before any model is evaluated** to prevent selection bias.
2. **Generation:** Greedy decoding, max 256 new tokens. No sampling — removes stochasticity as a confound.
3. **Judge:** GPT-4o via OpenAI API, scoring each completion on three dimensions (1–5):
   - **Coherence** — on-topic, grammatically well-formed
   - **Factual accuracy** — for factual prompts, verifiable claims
   - **Instruction adherence** — prompt structure followed
4. **Aggregation:** Mean score per dimension per model variant; 95% CI via bootstrap resampling (n=1000 resamples over prompts).

**Judge prompt template** (store in \`evals/judge_prompt.txt\`):
```
You are a rigorous evaluator. Rate the following model completion on three dimensions
(Coherence, Factual Accuracy, Instruction Adherence) each on a scale of 1–5.
Output ONLY a JSON object: {"coherence": N, "factual": N, "adherence": N}.

Prompt: {prompt}
Completion: {completion}
```

**Cost:** ~$4–6 per model variant (200 prompts × ~300 tokens). Run once on `best_model.pt` only.

**Prerequisite:** `scripts/generate.py` must be implemented first (currently a placeholder).

**Implementation:** `evals/llm_judge.py`, invoked via `scripts/eval.py --task llm_judge`

---

### 1.4 Inference Efficiency Profiling

**Why:** `train.py` already logs training throughput (`tok_per_sec`, \`dt\`). What is **missing** is **inference**-time profiling — latency, peak memory at inference batch sizes, and the overhead ratio between router types. This is where the \`find_unused_parameters=False\` advantage of the metabolic router has a concrete inference-time complement.

| Metric | Methodology |
|---|---|
| **Inference throughput (tokens/sec)** | Time 100 forward passes at batch=1 and batch=32; warmup=10; report mean ± std |
| **Inference latency (ms/token)** | 1000/throughput; report P50 and P95 |
| **Peak GPU memory (inference)** | `torch.cuda.max_memory_allocated()` at batch=32, seq=1024 |
| **Router overhead ratio** | `latency(metabolic) / latency(standard)` — shows fatigue adds negligible overhead |

Note: training throughput is already in WandB. This task is inference-only benchmarking on a loaded checkpoint.

**Implementation:** `evals/efficiency.py`, invoked via `scripts/eval.py --task efficiency`

---

### 1.5 Safety / Truthfulness

**Why:** Routing specialisation could cause factual queries to be routed to experts not trained on factual content, producing confident but wrong outputs. TruthfulQA is the standard benchmark for this failure mode.

| Task | Dataset | Metric | Shot |
|---|---|---|---|
| Truthfulness | **TruthfulQA** (MC1) | Acc | 0-shot |
| Truthfulness | **TruthfulQA** (MC2) | Acc | 0-shot |

MC1 (single correct answer) is the primary metric. MC2 is secondary. Both available directly via `lm-evaluation-harness` — no separate implementation needed beyond §1.2.

---

### P0 — Blocking paper submission

- [ ] **`evals/perplexity.py`** — sliding-window BPB/PPL on WikiText-103 test and PTB. Load checkpoint via `CheckpointManager.load_checkpoint()`. Run single-GPU, `bf16`, stride=512.
- [ ] **`evals/lm_harness_runner.py`** — thin wrapper around `lm_eval.simple_evaluate()`. Accept checkpoint path, task list, output path. Serialize to JSON per §3.
- [ ] **`scripts/eval.py`** — main entrypoint dispatching to the above. Args: `--task`, `--checkpoint`, `--config`, `--output-dir`. Remove the current `NotImplementedError`.
- [ ] **`evals/results_schema.py`** — canonical result JSON schema (see §3).

### P1 — Required for strong submission

- [ ] **`scripts/generate.py`** — generation pipeline (greedy + sampling). Required by `llm_judge.py`. Args: `--checkpoint`, `--prompt-file`, `--output-file`, `--max-new-tokens`, `--strategy [greedy|sampling]`.
- [ ] **`evals/judge_prompts.jsonl`** — 200 fixed prompts (factual × 70, reasoning × 70, creative × 60). **Must be committed before any model is evaluated.**
- [ ] **`evals/llm_judge.py`** — GPT-4o judge pipeline. Load prompts, generate completions via `generate.py`, call OpenAI API, aggregate with bootstrap CI.
- [ ] **`evals/efficiency.py`** — inference throughput + peak memory. Single-GPU, no DDP.
- [ ] **WandB eval logging** — after each eval task, log all scalar results to WandB project `tmoe`, run name `eval/{experiment_name}/step{N}`. Use `wandb.Table` for MMLU per-subject breakdown.

### P2 — Strengthening

- [ ] **Multi-seed variance** — re-run metabolic_v4 and standard_v2 with seeds {42, 123, 456}. Report mean ± std on all P0 metrics. Single-seed results are unlikely to survive peer review at top venues.
- [ ] **Routing metric correlation** — scatter plot of `effective_experts` (already in WandB) vs. downstream task accuracy across checkpoint steps. Demonstrates routing quality causally predicts task performance — the strongest claim in the paper.
- [ ] **`evals/plot.py`** — generate all paper figures from JSON artifacts (see §3). No manual copy-paste of numbers.

---

### Result Schema

Every eval run writes to `outputs/{experiment_name}/eval/{task}.json`:

```json
{
  "experiment_name": "gptneo_125m_metabolic_v4",
  "checkpoint_step": 5000,
  "checkpoint_path": "outputs/.../checkpoint_step_5000.pt",
  "eval_timestamp": "2026-03-09T14:32:00Z",
  "git_commit": "<sha>",
  "task": "perplexity",
  "config": { ... },
  "results": {
    "wikitext103_bpb": 1.234,
    "wikitext103_ppl": 12.34,
    "ptb_ppl": 45.67
  },
  "metadata": {
    "dtype": "bfloat16",
    "stride": 512,
    "device": "cuda:0",
    "torch_version": "2.3.0"
  }
}
```

**Hard requirements:**
- `git_commit` is mandatory — results without a pinned commit are not reproducible.
- `config` embeds the full experiment YAML contents (not a path) — the JSON must be self-contained.
- All floats stored at full Python precision. Rounding happens only at paper display time.

### Figures (generated by `evals/plot.py` from JSON artifacts — no manual numbers)

| Figure | Source data |
|---|---|
| Fig 1: Gini trajectory, metabolic vs. standard | WandB `router/{layer}/routing_diversity_gini` |
| Fig 2: Effective experts over training | WandB `router/{layer}/effective_experts` |
| Fig 3: MI layer-depth specialisation gradient | WandB `specialization/{tracker}/conditional_entropy` per layer |
| Table 1: PPL + downstream accuracy at equal FLOPs/token | `eval/perplexity.json` + `eval/lm_harness.json` |
| Table 2: Inference efficiency (throughput, memory, overhead ratio) | `eval/efficiency.json` |

---

## Contributor Notes

- **Do not modify `scripts/train.py`** — eval runs on saved checkpoints only.
- **Do not re-implement metrics already in `RouterMetricsTracker` or `GlobalSpecializationTracker`** — import and reuse them if needed.
- All eval scripts must run on a **single A100/H100** without multi-GPU. Use `CheckpointManager.load_checkpoint()` for checkpoint loading (handles DDP/FSDP unwrapping correctly).
- Test against the `smoketest.yaml` checkpoint before running on full checkpoints.
- Open a draft PR referencing this issue before writing code — describe your intended module structure so interfaces can be aligned before implementation.
